docs/adr/010_device_request_queue.md

```diff
@@ -0,0 +1,164 @@
+# Device Request Queue
+
+- **Feature**: [#1480]
+- **Author(s)**: Atif Aziz
+- **Status**: Proposed
+
+## Context
+
+LoRa devices, which can reach multiple LNS, pose an interesting challenge for
+Edge Hub since Edge Hub wasn't designed for devices competing for gateways.
+When a device connection is made to IoT Hub, Edge Hub continues to maintain
+that connection in the open state (for network resiliency) until explicitly
+closed. Moreover, IoT Hub allows maintining only a single connection per
+device. When another is opened, the existing one closed. If a device can reach
+multiple LNS then Edge Hub and IoT Hub begin an aggressive ping-pong of
+connections being opened and closed. In other words, as each LNS or Edge Hub
+tries to open a connection for the same device, IoT Hub closes the connection
+for the other. But then the other tries to re-open the connection and the
+ping-pong continues. This problem needs addressing to avoid scalability
+limitations.
+
+The ping-pong can be pevented by electing a leader LNS for a device and
+announcing the decision to others using a C2D message. An LNS receiving such a
+message will then close that connection so that Edge Hub does't continue to
+maintain the connection in the open state through retires. The handling of the
+message needs to know:
+
+1. It can close the connection right now, without interfering with a currently
+   running operation (like a message delivery where the message could be
+   potentially dropped).
+
+1. There are no operations starting around the same time that would cause it
+   to be opened again.
+
+The following cases can trigger operations from different threads:
+
+- An OTAA join message; `DeviceJoinLoader` eventually loads the twins and
+  stores them back up.
+
+- A data message; `DeviceLoaderSynchronizer` loads the twins for multiple
+  devices based on the `DevAddr` and adds them to the cache.
+
+- C2D message that needs to be delivered can trigger a load if the device is
+  not in the cache.
+
+- Cache refreshes can happen on a background thread for a device, causing the
+  need for a connection.
+
+**Note**: We should consider the option to keep building on top of the connection
```

Comment:
this was more a comment for you @atifaziz that we may want to list that as an option and consider it for the solution?